File: pset1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

bool isLuhnAlgVerified(char *nb)
{
    char c[20][20];
    char tmp[10];
    int total = 0;
    int subtotal = 0;
    int i = 0;
    int x = 0;
    while (i < strlen(nb))
    {
        if (i % 2 != 0)
        {
            sprintf(tmp, "%i", ((nb[i] - '0') * 2));
            if (atoi(tmp) > 9)
            {
                strcpy(c[x], "1");
                x++;
            }
            sprintf(tmp, "%i", (atoi(tmp) % 10));
            strcpy(c[x], tmp);
            x++;
        }
        else
            total = total + (nb[i] - '0');
        i++;
    }
    strcpy(c[x], "\0");
    i = 0;
    while (strcmp(c[i], "\0") != 0)
    {
        subtotal = subtotal + atoi(c[i]);
        i++;
    }
    if ((total + subtotal) % 10 == 0)
        return true;
    return false;
}

bool isValid(char *nb)
{
    int len = strlen(nb);
    if ((len == 15 || len == 16 || len == 13) &&
        isLuhnAlgVerified(nb))
        return true;
    return false;
}

bool isMastercard(char *nb)
{
    int len = strlen(nb);
    if (len == 16 && nb[0] == '5' &&
        (nb[1] == '1' || nb[1] == '2' || nb[1] == '3' || nb[1] == '4' || nb[1] == '5'))
        return true;
    return false;
}

bool isVisa(char *nb)
{
    int len = strlen(nb);
    if ((len != 16 && len != 13) && nb[0] != '4')
        return false;
    return true;
}

bool isAmex(char *nb)
{
    int len = strlen(nb);
    if (len == 15 && nb[0] == '3' &&
        (nb[1] == '4' || nb[1] == '7'))
        return true;
    return false;
}

bool checkCard(char *nb)
{
    if (isMastercard(nb))
    {
        printf("MASTERCARD\n");
        return true;
    }
    else if (isVisa(nb))
    {
        printf("VISA\n");
        return true;
    }
    else if (isAmex(nb))
    {
        printf("AMEX\n");
        return true;
    }
    return false;
}

int main(void)
{
    char nb[100] = "";
    long long number = 0;

    printf("Number : ");
    number = get_long_long();
    sprintf(nb, "%lld", number);

    if (!isValid(nb))
        printf("INVALID\n");
    else
        checkCard(nb);
    return 0;
}

File: pset6/credit.py
from cs50 import *


def isLuhnAlgVerified(nb):
    c = list("3782822463100056")
    total = 0
    subtotal = 0
    i = 0
    x = 0
    l = len(nb)
    while i < l:
        if i % 2 != 0:
            tmp = ((int(nb[i]) - int('0')) * 2)
            if tmp > 9:
                c[x] = "1"
                x += 1
            tmp = tmp % 10
            c[x] = str(tmp)
            x += 1
        else:
            total = total + (int(nb[i]) - int('0'))
        i += 1
    c[x] = "\0"
    i = 0
    while c[i] != "\0":
        subtotal = subtotal + int(c[i])
        i += 1
    if (total + subtotal) % 10 == 0:
        return True
    return False


def isValid(nb):
    l = len(nb)
    if ((l == 15 or len == 16 or len == 13) and
            isLuhnAlgVerified(nb)):
        return True
    return False


def isMasterCard(nb):
    l = len(nb)
    if (l == 16 and nb[0] == '5' and
            (nb[1] == '1' or nb[1] == '2' or nb[1] == '3' or nb[1] == '4' or nb[1] == '5')):
        return True
    return False


def isVisa(nb):
    l = len(nb)
    if (l != 16 or l != 13) and nb[0] != '4':
        return False
    return True


def isAmex(nb):
    l = len(nb)
    if (l == 15 and nb[0] == '3' and
            (nb[1] == '4' or nb[1] == '7')):
        return True
    return False


def checkCard(nb):
    if isMasterCard(nb):
        print("MASTERCARD")
    elif isVisa(nb):
        print("VISA")
    elif isAmex(nb):
        print("AMEX")


def main():
    print("Number : ", end='')
    nb = get_string()

    if not isValid(nb):
        print("INVALID")
    else:
        checkCard(nb)


if __name__ == "__main__":
    main()
